jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/FragmentingFlusher.java

```diff
@@ -0,0 +1,205 @@
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.core.FrameHandler.Configuration;
+import org.eclipse.jetty.websocket.core.OpCode;
+
+/**
+ * Used to split large data frames into multiple frames below the maxFrameSize.
+ */
+public abstract class FragmentingFlusher
+{
+    private static final Logger LOG = Log.getLogger(FragmentingFlusher.class);
+
+    private final Queue<FrameEntry> entries = new ArrayDeque<>();
+    private final IteratingCallback flusher = new Flusher();
+    private final Configuration configuration;
+    private boolean canEnqueue = true;
+
+    public FragmentingFlusher(Configuration configuration)
+    {
+        this.configuration = configuration;
+    }
+
+    abstract void forwardFrame(Frame frame, Callback callback, boolean batch);
+
+    public void sendFrame(Frame frame, Callback callback, boolean batch)
+    {
+        boolean enqueued = false;
+        synchronized (this)
+        {
+            if (canEnqueue)
+                enqueued = entries.offer(new FrameEntry(frame, callback, batch));
+        }
+
+        if (enqueued)
+            flusher.iterate();
+        else
+            notifyCallbackFailure(callback, new ClosedChannelException());
+    }
+
+    protected void onFailure(Throwable t)
+    {
+        synchronized (this)
+        {
+            canEnqueue = false;
+        }
+
+        for (FrameEntry entry : entries)
+            notifyCallbackFailure(entry.callback, t);
+        entries.clear();
+    }
+
+    private FrameEntry pollEntry()
+    {
+        synchronized (this)
+        {
+            return entries.poll();
+        }
+    }
+
+    private class Flusher extends IteratingCallback implements Callback
+    {
+        private FrameEntry current;
+        private boolean finished = true;
+
+        @Override
+        protected Action process()
+        {
+            if (finished)
+            {
+                current = pollEntry();
+                LOG.debug("Processing {}", current);
+                if (current == null)
+                    return Action.IDLE;
+
+                finished = false;
+                fragment(current, true);
+            }
+            else
+            {
+                fragment(current, false);
+            }
+            return Action.SCHEDULED;
+        }
+
+        private void fragment(FrameEntry entry, boolean first)
+        {
+            Frame frame = entry.frame;
+            ByteBuffer payload = frame.getPayload();
+            int remaining = payload.remaining();
+            long maxFrameSize = configuration.getMaxFrameSize();
+            int fragmentSize = (int)Math.min(remaining, maxFrameSize);
+
+            boolean continuation = (frame.getOpCode() == OpCode.CONTINUATION) || !first;
+            Frame fragment = new Frame(continuation ? OpCode.CONTINUATION : frame.getOpCode());
```

Comment:
Isn't there a common case where the passed frame is the first and smaller than maxFrame, so we can just forward it directly without ever creating a fragment Frame?